lib/perfdata/elasticsearchdatastreamwriter.cpp

```diff
@@ -0,0 +1,832 @@
+			metric->Set("warn", pdv->GetWarn());
+		if (!pdv->GetCrit().IsEmpty() && GetEnableSendThresholds())
+			metric->Set("crit", pdv->GetCrit());
+
+		pd_fields->Set(pdv->GetLabel(), metric);
+	}
+
+	return pd_fields;
+}
+
+// user-defined tags, as specified in the ECS specification: https://www.elastic.co/docs/reference/ecs/ecs-base#field-tags
+static Array::Ptr ExtractTemplateTags(const MacroProcessor::ResolverList& resolvers,
+	const Array::Ptr& tag_tmpl, const Checkable::Ptr& checkable, const CheckResult::Ptr& cr)
+{
+	if (tag_tmpl == nullptr) {
+		return nullptr;
+	}
+
+	ObjectLock olock(tag_tmpl);
+	Array::Ptr tags = new Array();
+	for (const String tag : tag_tmpl) {
+		String missingMacro;
+		Value value = MacroProcessor::ResolveMacros(tag, resolvers, cr, &missingMacro);
+		if (!missingMacro.IsEmpty()) {
+			Log(LogDebug, "ElasticsearchDatastreamWriter")
+				<< "Missing macro for tag: " << tag
+				<< ". Missing: " << missingMacro
+				<< " for checkable '" << checkable->GetName() << "'. Skipping.";
+			continue;
+		}
+		tags->Add(value);
+	}
+
+	return tags;
+}
+
+// user-defined labels, as specified in the ECS specification: https://www.elastic.co/docs/reference/ecs/ecs-base#field-labels
+static Dictionary::Ptr ExtractTemplateLabels(const MacroProcessor::ResolverList& resolvers,
+	const Dictionary::Ptr& labels_tmpl, const Checkable::Ptr& checkable, const CheckResult::Ptr& cr)
+{
+	if (labels_tmpl == nullptr) {
+		return nullptr;
+	}
+
+	ObjectLock olock(labels_tmpl);
+	Dictionary::Ptr labels = new Dictionary();
+	for (const Dictionary::Pair& label_kv : labels_tmpl) {
+		String missingMacro;
+		Value value = MacroProcessor::ResolveMacros(label_kv.second, resolvers, cr, &missingMacro);
+		if (!missingMacro.IsEmpty()) {
+			Log(LogDebug, "ElasticsearchDatastreamWriter")
+				<< "Missing macro for label: " << label_kv.first
+				<< "Label: " << label_kv.second
+				<< ". Missing: " << missingMacro
+				<< " for checkable '" << checkable->GetName() << "'. Skipping.";
+			continue;
+		}
+		labels->Set(label_kv.first, value);
+	}
+
+	return labels;
+}
+
+String ElasticsearchDatastreamWriter::ExtractDatastreamNamespace(const MacroProcessor::ResolverList& resolvers,
+	const Checkable::Ptr& checkable, const CheckResult::Ptr& cr)
+{
+	String namespace_tmpl = GetDatastreamNamespace();
+
+	String missingMacro;
+	Value value = MacroProcessor::ResolveMacros(namespace_tmpl, resolvers, cr, &missingMacro);
+	if (!missingMacro.IsEmpty() || value.IsEmpty()) {
+		Log(LogDebug, "ElasticsearchDatastreamWriter")
+			<< "Missing value for namespace. Missing: " << missingMacro
+			<< " for checkable '" << checkable->GetName() << "'. Skipping.";
+		return String("default");
+	}
+
+	return value;
+}
+
+bool ElasticsearchDatastreamWriter::Filter(const Checkable::Ptr& checkable, const CheckResult::Ptr& cr) {
+	if (!m_CompiledFilter)
+		return true;
+
+	Host::Ptr host;
+	Service::Ptr service;
+	tie(host, service) = GetHostService(checkable);
```

Comment:
You can use structured bindings, like so:

```suggestion
	auto [host, service] = GetHostService(checkable);
```